File: angry_cows.cpp
/*
USACO 2016 January Contest, Bronze
Problem 2. Angry Cows

Link to Problem: 
http://www.usaco.org/index.php?page=viewproblem2&cpid=592

*/



#include <iostream>
#include <cstdio>
#include <vector>
#include <algorithm>
using std::string;
using std::cout;
using std::cin;
using std::endl;
using std::sort;
using std::copy;
using std::vector;
using std::max;


int n;
vector<int> blocs;



int number_destroyed  (int cow_loc, int direction) {

    int blast_radius = 1;
    int before = cow_loc;

    //foward direction
    while (true) {
        int next = before;
        //find the furthest out haybale that explodes
        while (
            next + direction >= 0
            && next + direction < n
            && abs(blocs[next + direction] - blocs[before]) <= blast_radius) 
            {next += direction;}

        if (next==before) {
            break;
        }

        before = next;
        blast_radius++;
    }
    return abs(before - cow_loc);

}

int main()
{
    //uncomment to submit
    // freopen("angry.in", "r", stdin);
    // freopen("angry.out", "w", stdout);
    
    //get the stuff
    cin >> n;
    blocs.resize(n);
    for (int i = 0; i < n; i++) {
        cin >> blocs[i];
    }

    sort(blocs.begin(), blocs.end());

    int highest_destroyed = 0;

    for (int f = 0;f <= n; f++) {
        //for each location, count the number destroyed
        //compare with current highest
        int x = number_destroyed(f, -1) + 1 + number_destroyed(f, 1);
        highest_destroyed = max(highest_destroyed, x);

    }

    cout << highest_destroyed << endl;

}

File: USACO/bronze/angry_cows.py
"""
PYTHON SOLUTION

USACO 2016 January Contest, Bronze
Problem 2. Angry Cows

link to problem: 
http://www.usaco.org/index.php?page=viewproblem2&cpid=592
"""

# import sys
# sys.stdin = open("angry.in", "r")
# sys.stdout = open("angry.out", "w")


n = int(input())
locs = []
for _ in range(n):
    locs.append(int(input()))


max_d = 0

for loc in locs:
    time = 1
    all_d = 1
    locs_copy = locs.copy()
    locs_copy.remove(loc)
    new_d = True
    upper_bound,lower_bound = loc, loc


    while new_d:

        exploding = []
        

        for loc in locs_copy:
            if loc <= (upper_bound + time) and loc >= (lower_bound - time):
                exploding.append(loc)

        if len(exploding) == 0:
            new_d = False
            max_d = max(max_d, all_d)

        else:
            all_d += len(exploding)
            lower_bound = min(exploding)
            upper_bound = max(exploding)
            for item in exploding:
                locs_copy.remove(item)
            time += 1

print(max_d)